File: cli_terminal/output/collapse.py
"""Region collapsing for output content."""

from dataclasses import dataclass
from typing import List, Optional


@dataclass
class CollapseRegion:
    """A collapsible region."""
    start: int
    end: int
    header: str
    collapsed: bool = False


class CollapseManager:
    """Manages collapsible regions."""

    def __init__(self):
        self._regions: List[CollapseRegion] = []
        self._next_id: int = 0

    def add_region(self, start: int, end: int, header: str = "") -> int:
        """Add a collapsible region."""
        region_id = self._next_id
        self._next_id += 1
        self._regions.append(CollapseRegion(
            start=start,
            end=end,
            header=header or f"Region {region_id}",
        ))
        return region_id

    def toggle(self, index: int) -> bool:
        """Toggle collapse state of region."""
        if 0 <= index < len(self._regions):
            self._regions[index].collapsed = not self._regions[index].collapsed
            return self._regions[index].collapsed
        return False

    def is_collapsed(self, line_index: int) -> Optional[int]:
        """Check if line is in collapsed region, return region index."""
        for i, region in enumerate(self._regions):
            if region.collapsed and region.start <= line_index < region.end:
                return i
        return None

    def get_visible_lines(self, lines: List[str]) -> List[str]:
        """Get list of visible lines (excluding collapsed)."""
        result = []
        for i, line in enumerate(lines):
            collapsed = self.is_collapsed(i)
            if collapsed is None:
                result.append(line)
            elif i == self._regions[collapsed].start:
                result.append(f"[+ {self._regions[collapsed].header}]")
        return result

File: src/output/collapse.rs
/// A collapsible region within the output.
#[derive(Debug)]
pub struct CollapseRegion {
    pub start: usize,
    pub end: usize,
    pub header: String,
    pub collapsed: bool,
}

pub struct CollapseManager {
    regions: Vec<CollapseRegion>,
    next_id: usize,
}

impl Default for CollapseManager {
    fn default() -> Self {
        Self::new()
    }
}

impl CollapseManager {
    pub fn new() -> Self {
        Self {
            regions: Vec::new(),
            next_id: 0,
        }
    }

    /// Add a new collapsible region. Returns its ID.
    pub fn add_region(&mut self, start: usize, end: usize, header: &str) -> usize {
        let id = self.next_id;
        self.next_id += 1;
        self.regions.push(CollapseRegion {
            start,
            end,
            header: header.to_string(),
            collapsed: false,
        });
        id
    }

    /// Toggle the collapse state of region `id`. Returns the new state.
    pub fn toggle(&mut self, id: usize) -> Option<bool> {
        let region = self.regions.get_mut(id)?;
        region.collapsed = !region.collapsed;
        Some(region.collapsed)
    }

    /// Check if `line_idx` falls inside a collapsed region.
    pub fn is_hidden(&self, line_idx: usize) -> bool {
        self.regions
            .iter()
            .any(|r| r.collapsed && line_idx > r.start && line_idx < r.end)
    }

    /// Return the region ID whose header is at `line_idx`, if any.
    pub fn region_at(&self, line_idx: usize) -> Option<usize> {
        self.regions
            .iter()
            .position(|r| r.start == line_idx && r.collapsed)
    }

    /// Get the header text for a collapsed region at `line_idx`.
    pub fn collapsed_header_at(&self, line_idx: usize) -> Option<&str> {
        self.regions
            .iter()
            .find(|r| r.start == line_idx && r.collapsed)
            .map(|r| r.header.as_str())
    }

    pub fn regions(&self) -> &[CollapseRegion] {
        &self.regions
    }

    /// Toggle the most recently added region. Returns the new state.
    pub fn toggle_last(&mut self) -> Option<bool> {
        if self.regions.is_empty() {
            return None;
        }
        let last_idx = self.regions.len() - 1;
        self.regions[last_idx].collapsed = !self.regions[last_idx].collapsed;
        Some(self.regions[last_idx].collapsed)
    }
}
